File: kirun-js/__tests__/engine/function/system/array/CompareTest.ts
import { AbstractArrayFunction } from '../../../../../src/engine/function/system/array/AbstractArrayFunction';
import { Compare } from '../../../../../src/engine/function/system/array/Compare';
import { FunctionOutput } from '../../../../../src/engine/model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../../../src/engine/util/MapUtil';

test('Compare Test 1', async () => {
    let compare: Compare = new Compare();

    let source: any[] = [2, 2, 3, 4, 5];

    let find: any[] = [2, 2, 2, 3, 4, 5];

    expect(compare.compare(source, 0, 2, find, 1, 3)).toBe(0);

    find = [2, 2, 3, 4, 5];

    expect(compare.compare(source, 0, source.length, find, 0, find.length)).toBe(0);

    source = [true, true];

    find = [true, null];

    expect(compare.compare(source, 0, source.length, find, 0, find.length)).toBe(1);
});

test('Compare Test 2', async () => {
    let compare: Compare = new Compare();

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters();

    let source: any[] = [4, 5];

    let find: any[] = [4, 6];

    fep.setArguments(
        MapUtil.of(
            AbstractArrayFunction.PARAMETER_ARRAY_SOURCE.getParameterName(),
            source,
            Compare.PARAMETER_ARRAY_FIND.getParameterName(),
            find,
        ),
    );

    let fo: FunctionOutput = await compare.execute(fep);

    expect(fo.allResults()[0].getResult().get(AbstractArrayFunction.EVENT_RESULT_NAME)).toBe(5 - 6);
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/CompareTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class CompareTest {

	@Test
	void testInternalExecute() {

		Compare compare = new Compare();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

		JsonArray source = new JsonArray();
		source.add(4);
		source.add(5);

		JsonArray find = new JsonArray();
		find.add(4);
		find.add(6);

		fep.setArguments(Map.of(Compare.PARAMETER_ARRAY_SOURCE.getParameterName(), source,
				Compare.PARAMETER_ARRAY_FIND.getParameterName(), find));

		StepVerifier.create(compare.execute(fep))
				.expectNextMatches(result -> result.next().getResult()
						.get(Equals.EVENT_RESULT_NAME)
						.equals(new JsonPrimitive(Integer.compare(5, 6))))
				.verifyComplete();
	}

	@Test
	void testCompare() {

		Compare compare = new Compare();

		JsonElement[] source = new JsonElement[] { new JsonPrimitive(2), new JsonPrimitive(2), new JsonPrimitive(3),
				new JsonPrimitive(4), new JsonPrimitive(5) };

		JsonElement[] find = new JsonElement[] { new JsonPrimitive(2d), new JsonPrimitive(2d), new JsonPrimitive(2d),
				new JsonPrimitive(3d), new JsonPrimitive(4d), new JsonPrimitive(5d) };

		assertEquals(0, compare.compare(source, 0, 2, find, 1, 3));

		find = new JsonElement[] { new JsonPrimitive(2d), new JsonPrimitive(2d), new JsonPrimitive(3d),
				new JsonPrimitive(4d), new JsonPrimitive(5d) };

		assertEquals(0, compare.compare(source, 0, source.length, find, 0, find.length));

		source = new JsonElement[] { new JsonPrimitive(true), new JsonPrimitive(true) };

		find = new JsonElement[] { new JsonPrimitive(true), null };

		assertEquals(1, compare.compare(source, 0, source.length, find, 0, find.length));

		source = new JsonElement[] { new JsonPrimitive(true), JsonNull.INSTANCE };

		assertEquals(0, compare.compare(source, 0, source.length, find, 0, find.length));
	}

	@Test
	void testCompareTo() {

		Compare compare = new Compare();

		assertEquals(2 - 3, compare.compareTo(2, 3));
		assertEquals(2.0d - 3, compare.compareTo(2.0d, 3));
		assertEquals(Float.compare(30f, 3f), compare.compareTo(30, 3.0f));
	}

}
